src/general/world_effects/PhotosynthesisProductionEffect.cs

```diff
@@ -114,27 +113,48 @@ private void ApplyCompoundsAddition()
                     {
                         if (output.Key.ID is Compound.Oxygen)
                         {
-                            oxygenBalance += output.Value * outputModifier * effectiveSpeed * species.Value;
+                            oxygenOut += output.Value * outputModifier * effectiveSpeed * species.Value;
                         }
                         else if (output.Key.ID is Compound.Carbondioxide)
                         {
-                            co2Balance += output.Value * outputModifier * effectiveSpeed * species.Value;
+                            co2Out += output.Value * outputModifier * effectiveSpeed * species.Value;
                         }
                     }
                 }
             }
 
-            // Scale the balances to make the changes less drastic
-            oxygenBalance *= modifier;
-            co2Balance *= modifier;
+            float oxygenTarget;
+            float co2Target;
+
+            patch.Biome.TryGetCompound(Compound.Oxygen, CompoundAmountType.Biome, out var existingOxygen);
+            patch.Biome.TryGetCompound(Compound.Carbondioxide, CompoundAmountType.Biome, out var existingCo2);
+
+            float total = existingOxygen.Ambient + existingCo2.Ambient;
 
-            changesToApply.Clear();
+            if (oxygenOut == 0)
+            {
+                if (co2Out == 0)
+                {
+                    // in the rare event we aren't making either compound, do nothing
```

Comment:
But now consuming the compounds also doesn't do anything? I think this is a wrong conclusion to make in the code.

Reply:
You are correct. If microbes had only bio processes that destroyed O2 or CO2 without producing anything in return, this code doesn't account for that. Right now the PhotosynthesisProductionEffect class is dependent on the special case of conservation of mass (and that CO2 and O2 are only being converted into each other)

Reply:
The code shouldn't depend on that then as someone will make a mod eventually that breaks the laws of physics and will get confused why it doesn't do things correctly.

Reply:
alright, I've modified the code to handle matter sinks, and while a matter generator will result in absurd numbers, I've protected against divide-by-zero exceptions.